nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,510 @@
+            '';
+            default = [ ];
+            example = [
+              "/home/*/.cache"
+              "/nix"
+            ];
+          };
+
+          doInit = mkOption {
+            type = types.bool;
+            description = ''
+              Run <command>borg init</command> if the
+              specified <option>repo</option> does not exist.
+              You should set this to <literal>false</literal>
+              if the repository is located on an external drive
+              that might not always be mounted.
+            '';
+            default = true;
+          };
+
+          prune.keep = mkOption {
+            # Specifying e.g. `prune.keep.yearly = -1`
+            # means there is no limit of yearly archives to keep
+            # The regex is for use with e.g. --keep-within 1y
+            type = with types; attrsOf (either int (strMatching "[[:digit:]]+[Hdwmy]"));
+            description = ''
+              Prune a repository by deleting all archives
+              not matching any of the specified retention options.
+            '';
+            default = { };
+            example = literalExample ''
+              {
+                within = "1d"; # Keep all archives from the last day
+                daily = 7;
+                weekly = 4;
+                monthly = -1; # Keep at least one archive for each month
+              }
+            '';
+          };
+
+          environment = mkOption {
+            type = with types; attrsOf str;
+            description = ''
+              Environment variables passed to the backup script.
+              You can for example specify which SSH key to use.
+            '';
+            default = { };
+            example = { BORG_RSH = "ssh -i /path/to/key"; };
+          };
+
+          preHook = mkOption {
+            type = types.lines;
+            description = ''
+              Shell commands to run before the backup.
+              This can for example be used to mount file systems.
+            '';
+            default = "";
+            example = ''
+              # In case you need to know the exact archive name
+              archiveName="foobar-$(date +%Y%m%dT%H%M%S)"
+            '';
+          };
+
+          postInit = mkOption {
+            type = types.lines;
+            description = ''
+              Shell commands to run after
+              <command>borg init</command>.
+            '';
+            default = "";
+          };
+
+          postCreate = mkOption {
+            type = types.lines;
+            description = ''
+              Shell commands to run after
+              <command>borg create</command>.
+            '';
+            default = "";
+            example = ''
+              # In case you want to be able to distinguish failed backups
+              # (and those during which a warning occurred)
+              borg rename "::$archiveName" "$archiveName-success"
```

Comment:
Is the script run with "set -e"? If so, that means postCreate will not be called on error. What I want to do is have the archive named "*-unsuccessful" in case of error, and that seems not possible with postCreate. What I could do is use postHook, but we'd need to document that the exit status is available in the shell variable "$exit_status" (and update the example).

Reply:
Yes it is run with the `-e` option:
https://github.com/NixOS/nixpkgs/blob/b7e51bf4ffb57197639c8ab3c7627973538910a6/nixos/modules/system/boot/systemd.nix#L258
However, the trap already has the effect of exiting on error so the `-e` option would not be necessary.

Yes, it is possible. You can name the archive `foobar-unsuccessful` in `preHook` and remove the suffix in `postCreate`. I did not use this in the example because it involves the two options and therefore two examples.

But thinking about it, we might add a boolean option like `tagFailure` which does the following:
- The archive name used for `borg create` will have an `-unsuccessful` or `-failure` suffix (whatever you like more).
- After `borg create`, the name of the last archive is found out using `borg list`. Because this will only be done if `borg create` was successful, the found archive will indeed be the one created during the current run of the backup job.
- Remove the suffix from the found archive.

Regarding the exit status, I will add a line that the exit status it available in `$?`. But I don't think that's very helpful because you might want to know *which* part of the script failed. Maybe I should just copy you `$create_ret` etc. stuff. What do you think?

Reply:
I implemented an option `appendFailedSuffix` which does exactly what I described above.

Now only the return code variables should be missing compared to your implementation. Any thoughts about that, @bjornfor?

Reply:
With appendFailedSuffix, I don't think I need to know the exit status in the postHook. If I want to know about overall service, I can use systemd.services.*.onFailure. Thanks!

Btw, the exit status won't be available in $? for the postHook, because the command that saves the status, "$exit_status=$?", sets the status to 0.

Reply:
One final thing I am slightly worried about: Because the lock is released between `borg create` and `borg rename`, another archive might have been created. I don't know a way to keep the lock during multiple operations but I think it is highly unlikely that this causes problems.

Reply:
I'm slightly worried about that too. That's why I personally stopped using borg variable placeholders and took ownership of the archive name in the NixOS module.

Reply:
I think the only real possibility would be if several clients regularly push to the same repo, with big --lock-wait N option. But in that case, I would guess it's pretty likely to hit an issue, over time.

Reply:
How about `borg create --json ... | jq .archive.name`

Reply:
That seems to fix the race, but now the output format is changed for everyone. I regularly look at the human readable --stats output. That won't work with this change. (I also expected this to silence all log output from borg, but things from --list --verbose goes to stderr it seems.)

Reply:
That's true :thinking: 
But you can now do something like
```
postCreate = ''
  echo $json | jq --compact-output  .archive.stats
'';
```
I will add something along those lines as an example.

Reply:
Btw, you can also get the same output as you would from `borg create` by using `borg info`.